clang/lib/AST/ASTContext.cpp

```diff
@@ -12726,11 +12726,15 @@ bool ASTContext::DeclMustBeEmitted(const Decl *D) {
 
   // Likewise, variables with tuple-like bindings are required if their
   // bindings have side-effects.
-  if (const auto *DD = dyn_cast<DecompositionDecl>(VD))
-    for (const auto *BD : DD->bindings())
-      if (const auto *BindingVD = BD->getHoldingVar())
-        if (DeclMustBeEmitted(BindingVD))
-          return true;
+  if (const auto *DD = dyn_cast<DecompositionDecl>(VD)) {
+    bool BindingResult = false;
+    DD->VisitHoldingVars([&](VarDecl *BindingVD) {
```

Comment:
Instead of a 'Visit' here, I wonder if there is value with some sort of `llvm::any_of` on the collection instead.

Reply:
I just added a flat_binding_iterator to replace these Visit* functions. I forgot to use it here, but if a custom iterator is what you wanted, please consider it.

Reply:
I'd be a little surprised if there wasn't an existing iterator type that would work.  If not though, I'd vastly prefer removing both of the `Visit` functions and replacing them with iterators.

Reply:
I replaced the Visit functions with `flat_bindings` and reverted the places where I was using them. I will look to see if there is some existing iterator that you are referring to if it would be simple than `flat_binding_iterator` (which had some things I had to fix.)